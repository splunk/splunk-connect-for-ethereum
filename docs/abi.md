# ABI Decoding

Splunk Connect for Ethereum features decoding of function calls and event logs in Ethereum transactions based on [Solidity ABI](https://solidity.readthedocs.io/en/develop/abi-spec.html) information.

In addition to the binary data of transactions, the following information is extracted:

-   Function or event name
-   Parameter values and data types
-   Parameter names
-   Contract name

This information can be used to easily search for semantic aspects of smart contracts and it enables powerful and efficient aggregations and analytics.

## Supplying ABI Definitions

You can supply ABI definition files to ethlogger by specifying a directory in which ethlogger should search for them. This can be done by specifying the `--abi-dir` [CLI flag](./cli.md), or by providing the corresponding [environment variable](./cli.md#environment-variables) `ABI_DIR`. For more fine-grained control, you can configure the [ABI repository in ethlogger configuration file](./configuration.md#abirepository).

Each ABI file corresponds to a smart contract and contains all function and event signatures, and optionally a contract name and address at which the contract is deployed.

Matching ABI defintions to transactions follows the following priority order:

-   Contract address (if contract address from ABI file matches address from transaction)
-   Contract fingerprint (if contract fingerprint matches fingerprint derived from contract code)
-   Anonymous signature

### File Formats

Files in the ABI directory are expected to contain [ABI JSON](https://solidity.readthedocs.io/en/develop/abi-spec.html#abi-json), either as plain list of ABI items, or truffle build files.

#### List of ABI Items

ABI files may contain an array of ABI items. In this case the contract name is derived from the file name (by removing the file extension) and no contract address is associated.

**Example**:

```json
[
    {
        "type": "function",
        "name": "transfer",
        "inputs": [
            {
                "name": "recipient",
                "type": "address"
            },
            {
                "name": "amount",
                "type": "uint256"
            }
        ]
    },
    {
        "type": "event",
        "name": "TransferApproved",
        "inputs": [
            {
                "name": "recipient",
                "type": "address"
            },
            {
                "name": "amount",
                "type": "uint256"
            }
        ]
    }
]
```

#### Truffle Build Files

Truffle build files can have information about which network and contract address a contract has been deployed to. This information is used by ethlogger to fast-track matching an address it encounters on the blockchain to the ABI definition. Furthermore, truffle build files can specify a contract name; if not present the file name will be used as contract name. Files are expected to have a property `abi` containing an array of ABI items, all other information is optional.

```jsonc
{
    "contractName": "My Contract",
    "abi": [
        /* ... ABI items */
    ],
    "networks": {
        "1": {
            "address": "0x123..."
        }
    }
}
```

## Contract Fingerprinting

One challenge that can occur when decoding ABI information is overlapping signatures. Imagine 2 smart contracts both exposing a function with the signature `transfer(address,uint256)`. For one contract the second argument may refer to an amount of currency being transferred (eg. `transfer(to address, amount uint256)`) and for another contract the second argument refers to the ID of a non-fungible item (eg. `transfer(to address, tokenId uint256)`). Given the nature of ABI encoding, both functions end up having the same signature and it's ambiguous for the ABI decoder which one to choose, unless the smart contract can be associated with the ABI definition (if the ABI definition contains the contract address).

It would be nice, to be able to decode any instance of a smart contract, like an ERC20 and not having to enumerate all the address the contract has been deployed to.

An attempt to solve this problem is "contract fingerprinting". A fingerprint is a hash of all function and event signatures from a contract ABI. We can also derive the fingerprint from the EVM bytecode by extracting all signature hashes, finding matching signatures for them, and computing the fingerprint from it, and then comparing that against the fingerprint computed from the ABI defintion. There is still the possibility of invalid matches, especially for contracts that only have very few functions/events, but with multiple functions the chances are much lower.

Fingerprinting is enabled by default and can be disabled in the [ethlogger configuration](./configuration.md#abirepository).

## Anonymous ABI Decoding

For cases where we can find a match with between a deployed contract and a supplied ABI definition, but do have a match for the function or event signature hash, we can emit the reduced amount of information we can derive from the signature alone (the function/event name, the parameter data types and values, but not the parameter names or the contract name).

If anonymous ABI encoding is enabled (which it is by default), then this reduced amount of information is emitted. In addition to the ABIs supplied by the user, ethlogger ships with a standard set of function and event signatures that are compiled from from external sources:

-   https://github.com/MrLuit/evm
-   https://www.4byte.directory/

Compiled lists can be found here:

-   Function signatures: [function_signatures.txt](../data/function_signatures.txt)
-   Event signatures: [event_signatures.txt](../data/event_signatures.txt)

Those anonymous signature are loaded by default.
