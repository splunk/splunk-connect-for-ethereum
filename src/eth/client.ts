import { sleep } from '../utils/async';
import { createModuleDebug } from '../utils/debug';
import { checkError, createJsonRpcPayload, JsonRpcError, JsonRpcRequest } from './jsonrpc';
import { EthRequest } from './requests';
import { EthereumTransport } from './transport';

const { debug, error } = createModuleDebug('eth:client');

interface BatchReq<P extends any[] = any[], R = any> {
    request: EthRequest<P, R>;
    callback: (error: Error | null, result: R) => void;
}

export async function executeBatchRequest(batch: BatchReq[], transport: EthereumTransport) {
    debug('Processing batch of %d JSON RPC requests', batch.length);
    try {
        const items = new Map<number, BatchReq>();
        const reqs: JsonRpcRequest[] = [];
        for (const batchItem of batch) {
            const req = createJsonRpcPayload(batchItem.request.method, batchItem.request.params);
            reqs.push(req);
            items.set(req.id, batchItem);
        }
        const results = await transport.sendBatch(reqs);
        for (const result of results) {
            const batchItem = items.get(result.id);
            if (batchItem == null) {
                error(`Found unassociated batch item in batch response`);
                continue;
            }
            batchItem.callback(null, result);
            items.delete(result.id);
        }
        if (items.size > 0) {
            error(`Unprocessed batch request after receiving results`);
            for (const unprocessed of items.values()) {
                unprocessed.callback(new Error('Result missing from batch response'), null);
            }
        }
    } catch (e) {
        batch.forEach(({ callback }) => callback(e, null));
    }
}

export class EthereumClient {
    constructor(protected transport: EthereumTransport) {}

    async request<P extends any[], R>(req: EthRequest<P, R>): Promise<R> {
        const payload = createJsonRpcPayload(req.method, req.params);
        const res = await this.transport.send(payload);
        if (payload.id !== res.id) {
            throw new JsonRpcError(`JSON RPC Response ID mismatch. Expected ${payload.id} but got ${res.id}`);
        }
        checkError(res);
        return typeof req.response === 'function' ? req.response(res) : res.result;
    }

    async requestBatch<P extends any[], R>(reqs: EthRequest<P, R>[]): Promise<R[]> {
        const batchReqs: BatchReq[] = [];
        const promises = [];
        for (const request of reqs) {
            promises.push(
                new Promise<R>((resolve, reject) => {
                    batchReqs.push({
                        request,
                        callback: (e, result) => {
                            if (e) {
                                reject(e);
                            }
                            try {
                                checkError(result);
                                resolve(
                                    typeof request.response === 'function' ? request.response(result) : result.result
                                );
                            } catch (e) {
                                debug(`JSON RPC request method: %s failed`, request.method, e);
                                reject(e);
                            }
                        },
                    });
                })
            );
        }
        await sleep(0);
        await executeBatchRequest(batchReqs, this.transport);
        return await Promise.all(promises);
    }
}

export interface BatchedEthereumClientConfig {
    maxBatchTime: number;
    maxBatchSize: number;
}

export class BatchedEthereumClient extends EthereumClient {
    private config: BatchedEthereumClientConfig;
    private queue: BatchReq<any[], any>[] = [];
    private flushTimer: NodeJS.Timer | null = null;
    constructor(transport: EthereumTransport, config: BatchedEthereumClientConfig) {
        super(transport);
        this.config = config;
    }

    request<P extends any[], R>(
        req: EthRequest<P, R>,
        { immediate = false }: { immediate?: boolean } = {}
    ): Promise<R> {
        if (immediate) {
            return super.request(req);
        }

        return new Promise((resolve, reject) => {
            this.queue.push({
                request: req,
                callback: (e, result) => {
                    if (e) {
                        reject(e);
                    }
                    try {
                        checkError(result);
                        resolve(typeof req.response === 'function' ? req.response(result) : result.result);
                    } catch (e) {
                        debug(`JSON RPC request method: %s(%o) failed`, req.method, req.params, e);
                        reject(e);
                    }
                },
            });
            this.scheduleFlush();
        });
    }

    private processQueue = async () => {
        if (this.flushTimer != null) {
            clearTimeout(this.flushTimer);
            this.flushTimer = null;
        }
        if (this.queue.length === 0) {
            return;
        }
        const batch = this.queue;
        this.queue = [];
        return await executeBatchRequest(batch, this.transport);
    };

    private scheduleFlush() {
        if (this.queue.length >= this.config.maxBatchSize) {
            this.processQueue();
            return;
        }
        if (this.flushTimer == null) {
            this.flushTimer = setTimeout(this.processQueue, this.config.maxBatchTime);
        }
    }
}
