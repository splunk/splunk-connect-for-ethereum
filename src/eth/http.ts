import { default as HttpAgent, HttpOptions, HttpsAgent } from 'agentkeepalive';
import fetch from 'node-fetch';
import { createModuleDebug } from '../utils/debug';
import { isHttps } from '../utils/http';
import { isValidJsonRpcResponse, JsonRpcRequest, JsonRpcResponse } from './jsonrpc';
import { EthereumTransport } from './transport';
import { httpClientStats } from '../utils/stats';

const { debug, trace } = createModuleDebug('eth:http');

export interface HttpTransportConfig {
    url: string;
    timeout?: number;
    validateCertificate?: boolean;
    requestKeepAlive?: boolean;
    maxSockets?: number;
}

const CONFIG_DEFAULTS = {
    timeout: 60_000,
    validateCertificate: true,
    requestKeepAlive: true,
    maxSockets: 256,
};

const initialCounters = {
    requests: 0,
    batches: 0,
    sumBatchSize: 0,
    errors: 0,
};

export class HttpTransport implements EthereumTransport {
    private config: HttpTransportConfig & typeof CONFIG_DEFAULTS;
    private httpAgent: HttpAgent | HttpsAgent;
    private counters = { ...initialCounters };

    constructor(config: HttpTransportConfig) {
        this.config = { ...CONFIG_DEFAULTS, ...config };
        const baseAgentOptions: HttpOptions = {
            keepAlive: true,
            maxSockets: 256,
        };
        this.httpAgent = isHttps(this.config.url)
            ? new HttpsAgent({
                  ...baseAgentOptions,
                  rejectUnauthorized: this.config.validateCertificate,
              })
            : new HttpAgent(baseAgentOptions);
    }

    public async send(request: JsonRpcRequest): Promise<JsonRpcResponse> {
        debug('Sending JSON RPC request: %o', request.method);
        const result = await this.sendInternal(request);
        if (Array.isArray(result)) {
            throw new Error(`JSON RPC returned batch but expected single message`);
        }
        return result;
    }

    public async sendBatch(request: JsonRpcRequest[]): Promise<JsonRpcResponse[]> {
        debug('Sending JSON RPC batch containing %d requests', request.length);
        const result = await this.sendInternal(request);
        if (!Array.isArray(result)) {
            throw new Error(`JSON RPC returned single message, was expecting batch`);
        }
        return result;
    }

    public async sendInternal(
        request: JsonRpcRequest | JsonRpcRequest[]
    ): Promise<JsonRpcResponse | JsonRpcResponse[]> {
        const startTime = Date.now();
        trace(`Sending JSON RPC request over HTTP: %O`, request);
        const body = JSON.stringify(request);
        this.counters.requests++;
        if (Array.isArray(request)) {
            this.counters.batches++;
            this.counters.sumBatchSize += request.length;
        }
        try {
            const response = await fetch(this.config.url, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                    'Content-Length': String(body.length),
                },
                body,
                agent: this.httpAgent,
                timeout: this.config.timeout,
            });

            if (response.status < 200 || response.status > 299) {
                throw new Error(
                    `JSON RPC service ${this.config.url} responded with HTTP status ${response.status} (${response.statusText})`
                );
            }

            const data = await response.json();

            trace('Received JSON RPC response:\n%O', data);
            if (!isValidJsonRpcResponse(data)) {
                throw new Error('Invalid JSON RPC response');
            }

            debug('Completed JSON RPC request in %d ms', Date.now() - startTime);
            return data as JsonRpcResponse | JsonRpcResponse[];
        } catch (e) {
            this.counters.errors++;
            throw e;
        }
    }

    public get stats() {
        return {
            ...this.counters,
            httpClient: httpClientStats(this.httpAgent.getCurrentStatus()),
        };
    }

    public flushStats() {
        const stats = this.stats;
        this.counters = { ...initialCounters };
        return stats;
    }
}
