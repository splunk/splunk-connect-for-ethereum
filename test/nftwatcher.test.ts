import { join } from 'path';
import { ContractInfo } from '../src/abi/contract';
import { State } from '../src/state';
import { BatchedEthereumClient } from '../src/eth/client';
import { HttpTransport } from '../src/eth/http';
import { withRecorder } from '../src/eth/recorder';
import { suppressDebugLogging } from '../src/utils/debug';
import LRUCache from '../src/utils/lru';
import { TestOutput } from './testoutput';
import { MOCK_NODE_ADAPTER } from '../src/platforms/mock';
import { NFTWatcher } from '../lib/nftwatcher';

let logHandle: any;
beforeEach(() => {
    logHandle = suppressDebugLogging();
});
afterEach(() => {
    logHandle.restore();
});

test('nftwatcher', async () => {
    await withRecorder(
        new HttpTransport('https://dai.poa.network', {}),
        {
            name: 'xdai-nftwatcher',
            storageDir: join(__dirname, './fixtures/recorded'),
            replay: true,
        },
        async transport => {
            const ethClient = new BatchedEthereumClient(transport, { maxBatchSize: 100, maxBatchTime: 0 });
            const state = new State({
                path: join(__dirname, '../tmp/tmpcheckpoint.json'),
                saveInterval: 10000,
            });
            const checkpoint = state.getCheckpoint('main');
            checkpoint.setInitialBlockNumber(123);
            const output = new TestOutput();
            const contractInfoCache = new LRUCache<string, Promise<ContractInfo>>({ maxSize: 100 });
            const nftWatcher = new NFTWatcher({
                checkpoint,
                config: {
                    contractAddress: '0x22c1f6050e56d2876009903609a2cc3fef83b415',
                    enabled: true,
                    blocksMaxChunkSize: 1,
                    pollInterval: 1,
                    maxParallelChunks: 1,
                    startAt: 'latest',
                    retryWaitTime: 10,
                    logEthBalance: true,
                },
                ethClient,
                output,
                contractInfoCache,
                waitAfterFailure: 1,
                nodePlatform: MOCK_NODE_ADAPTER,
                collectRetrievalTime: false,
            });

            await nftWatcher.processChunk({ from: 19595071, to: 19595091 });

            expect(output.messages).toMatchSnapshot();
        }
    );
}, 15000);
