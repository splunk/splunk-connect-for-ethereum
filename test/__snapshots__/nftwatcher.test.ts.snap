// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`nftwatcher 1`] = `
Array [
  Object {
    "body": Object {
      "blockHash": "0x5986758215563cf2873d5e43d7ba700bb80df4244a8a86b10e12a2a5f5de3716",
      "blockNumber": 19595071,
      "contract": "0x22c1f6050e56d2876009903609a2cc3fef83b415",
      "errorTokenURI": undefined,
      "from": "0x0000000000000000000000000000000000000000",
      "fromEthBalance": "105987609.019647827235641897",
      "metadata": Object {
        "attributes": Array [
          Object {
            "trait_type": "startDate",
            "value": "14-Dec-2021",
          },
          Object {
            "trait_type": "endDate",
            "value": "14-Dec-2021",
          },
          Object {
            "trait_type": "virtualEvent",
            "value": "true",
          },
          Object {
            "trait_type": "city",
            "value": "",
          },
          Object {
            "trait_type": "country",
            "value": "",
          },
          Object {
            "trait_type": "eventURL",
            "value": "https://www.nba.com/warriors/news/curry-nba-3point-leader-20211214",
          },
        ],
        "description": "Golden State Warriors guard Stephen Curry became the NBA’s all-time 3-point leader with 2,974 career 3-pointers, hitting his second three of the night with 7:33 remaining in the first quarter of tonight’s (12.14.2021) contest at the New York Knicks to surpass Ray Allen’s career mark (2,973 threes). ",
        "external_url": "https://api.poap.xyz/metadata/18202/3122679",
        "home_url": "https://app.poap.xyz/token/3122679",
        "image_url": "https://assets.poap.xyz/stephen-curry-becomes-nbas-all-time-3-point-leader-2021-logo-1639544399986.png",
        "name": "Stephen Curry Becomes NBA’s All-Time 3-Point Leader",
        "tags": Array [
          "poap",
          "event",
        ],
        "year": 2021,
      },
      "rawTokenURI": "0x0000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000002b68747470733a2f2f6170692e706f61702e78797a2f6d657461646174612f31383230322f33313232363739000000000000000000000000000000000000000000",
      "retrievalTime": 0,
      "to": "0xe83aee0a19700705726d8cfb39dc6032f151a6c4",
      "toEthBalance": "0.000000000000000000",
      "tokenIndex": "0x00000000000000000000000000000000000000000000000000000000002fa5f7",
      "tokenURI": "https://api.poap.xyz/metadata/18202/3122679",
      "transactionHash": "0xb025c4d21b7857f4a33a9e36073b0459a1b05c1e3096a04cf3a7c4a46180c46c",
    },
    "time": 1639586690000,
    "type": "nft",
  },
  Object {
    "body": Object {
      "blockHash": "0x5986758215563cf2873d5e43d7ba700bb80df4244a8a86b10e12a2a5f5de3716",
      "blockNumber": 19595071,
      "contract": "0x22c1f6050e56d2876009903609a2cc3fef83b415",
      "errorTokenURI": undefined,
      "from": "0x0000000000000000000000000000000000000000",
      "fromEthBalance": "105987609.019647827235641897",
      "metadata": Object {
        "attributes": Array [
          Object {
            "trait_type": "startDate",
            "value": "13-Dec-2021",
          },
          Object {
            "trait_type": "endDate",
            "value": "20-Dec-2021",
          },
          Object {
            "trait_type": "virtualEvent",
            "value": "true",
          },
          Object {
            "trait_type": "city",
            "value": "",
          },
          Object {
            "trait_type": "country",
            "value": "",
          },
          Object {
            "trait_type": "eventURL",
            "value": "https://sandbox.poap.art/",
          },
        ],
        "description": "You were at the intersection of art and tech on the Sandbox weekly canvas week 50 of POAP.art ",
        "external_url": "https://api.poap.xyz/metadata/17998/3122680",
        "home_url": "https://app.poap.xyz/token/3122680",
        "image_url": "https://assets.poap.xyz/poapart-sandbox-weekly-canvas-week-50-2021-2021-logo-1639449332633.png",
        "name": "POAP.art Sandbox Weekly Canvas - Week 50 2021",
        "tags": Array [
          "poap",
          "event",
        ],
        "year": 2021,
      },
      "rawTokenURI": "0x0000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000002b68747470733a2f2f6170692e706f61702e78797a2f6d657461646174612f31373939382f33313232363830000000000000000000000000000000000000000000",
      "retrievalTime": 0,
      "to": "0x9c7d47cdb2d3ecc31f5d778c591f8846aad0a83d",
      "toEthBalance": "0.000000000000000000",
      "tokenIndex": "0x00000000000000000000000000000000000000000000000000000000002fa5f8",
      "tokenURI": "https://api.poap.xyz/metadata/17998/3122680",
      "transactionHash": "0xe7e54156cb6ed90c825756d00e5bf8f8298fec35c76aae373adfb1d4dbaa71a3",
    },
    "time": 1639586690000,
    "type": "nft",
  },
  Object {
    "body": Object {
      "blockHash": "0x759479e48c81163e15d4466a6851de12f763d84587f4f6e21796970df887e565",
      "blockNumber": 19595077,
      "contract": "0x22c1f6050e56d2876009903609a2cc3fef83b415",
      "errorTokenURI": undefined,
      "from": "0x0000000000000000000000000000000000000000",
      "fromEthBalance": "105987609.019647827235641897",
      "metadata": Object {
        "attributes": Array [
          Object {
            "trait_type": "startDate",
            "value": "15-Dec-2021",
          },
          Object {
            "trait_type": "endDate",
            "value": "15-Dec-2021",
          },
          Object {
            "trait_type": "virtualEvent",
            "value": "true",
          },
          Object {
            "trait_type": "city",
            "value": "HYVE Islands",
          },
          Object {
            "trait_type": "country",
            "value": "",
          },
          Object {
            "trait_type": "eventURL",
            "value": "",
          },
        ],
        "description": "This is a digital token for all individuals that participated in the HYVE Digital Christmas Event 2021 as a member of the Strategy & Empowerment Team.",
        "external_url": "https://api.poap.xyz/metadata/17351/3122681",
        "home_url": "https://app.poap.xyz/token/3122681",
        "image_url": "https://assets.poap.xyz/hyve-digital-christmas-event-2021-2021-logo-1639068367114.png",
        "name": "HYVE Digital Christmas Event 2021",
        "tags": Array [
          "poap",
          "event",
        ],
        "year": 2021,
      },
      "rawTokenURI": "0x0000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000002b68747470733a2f2f6170692e706f61702e78797a2f6d657461646174612f31373335312f33313232363831000000000000000000000000000000000000000000",
      "retrievalTime": 0,
      "to": "0xcdc8fbc1cd62335902664e31295c87e47f7453da",
      "toEthBalance": "0.000000000000000000",
      "tokenIndex": "0x00000000000000000000000000000000000000000000000000000000002fa5f9",
      "tokenURI": "https://api.poap.xyz/metadata/17351/3122681",
      "transactionHash": "0xdfc458bdec9b85959a6a93173e156acfa739147517ffa60f5340cc528a378d28",
    },
    "time": 1639586720000,
    "type": "nft",
  },
  Object {
    "body": Object {
      "blockHash": "0x980533e8b922aabe37fc9277586713e5c333688efe9e06982a50873c97c3e87a",
      "blockNumber": 19595078,
      "contract": "0x22c1f6050e56d2876009903609a2cc3fef83b415",
      "errorTokenURI": undefined,
      "from": "0x0000000000000000000000000000000000000000",
      "fromEthBalance": "105987609.019647827235641897",
      "metadata": Object {
        "attributes": Array [
          Object {
            "trait_type": "startDate",
            "value": "14-Dec-2021",
          },
          Object {
            "trait_type": "endDate",
            "value": "14-Dec-2021",
          },
          Object {
            "trait_type": "virtualEvent",
            "value": "true",
          },
          Object {
            "trait_type": "city",
            "value": "",
          },
          Object {
            "trait_type": "country",
            "value": "",
          },
          Object {
            "trait_type": "eventURL",
            "value": "https://potos.io",
          },
        ],
        "description": "Work like a Captain, play like a pirate! This be POAP #14 of POTOS 25 days of POAPS t' celebrate the Holidays. AAArrrghh",
        "external_url": "https://api.poap.xyz/metadata/16277/3122682",
        "home_url": "https://app.poap.xyz/token/3122682",
        "image_url": "https://assets.poap.xyz/potos-25-days-of-poaps-14-2021-logo-1638464481588.png",
        "name": "POTOS 25 Days of POAPS (#14)",
        "tags": Array [
          "poap",
          "event",
        ],
        "year": 2021,
      },
      "rawTokenURI": "0x0000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000002b68747470733a2f2f6170692e706f61702e78797a2f6d657461646174612f31363237372f33313232363832000000000000000000000000000000000000000000",
      "retrievalTime": 0,
      "to": "0xa086c3bbf195aa18bee2805d568072b8af58fbd6",
      "toEthBalance": "0.000000000000000000",
      "tokenIndex": "0x00000000000000000000000000000000000000000000000000000000002fa5fa",
      "tokenURI": "https://api.poap.xyz/metadata/16277/3122682",
      "transactionHash": "0x5801d99eeb69bd5f652d4ce5b913d6849b2756f02b2691185325ab07108a1201",
    },
    "time": 1639586725000,
    "type": "nft",
  },
  Object {
    "body": Object {
      "blockHash": "0xf7567389680a3d08f1ebb67156b14082570709e5efd2505efee955c9e7c3a8b9",
      "blockNumber": 19595079,
      "contract": "0x22c1f6050e56d2876009903609a2cc3fef83b415",
      "errorTokenURI": undefined,
      "from": "0x0000000000000000000000000000000000000000",
      "fromEthBalance": "105987609.019647827235641897",
      "metadata": Object {
        "attributes": Array [
          Object {
            "trait_type": "startDate",
            "value": "14-Dec-2021",
          },
          Object {
            "trait_type": "endDate",
            "value": "14-Dec-2021",
          },
          Object {
            "trait_type": "virtualEvent",
            "value": "true",
          },
          Object {
            "trait_type": "city",
            "value": "",
          },
          Object {
            "trait_type": "country",
            "value": "",
          },
          Object {
            "trait_type": "eventURL",
            "value": "",
          },
        ],
        "description": "This POAP commemorates attendance of the Dec. 14, 2021 Bullish Radio (hosted by @hunterleroyy) space, in which collectors @boredethguy and @newethguy discussed topics related to collecting, NFTs, art, and life.",
        "external_url": "https://api.poap.xyz/metadata/18199/3122683",
        "home_url": "https://app.poap.xyz/token/3122683",
        "image_url": "https://assets.poap.xyz/2021-12-14-bullish-radio-the-collectors-eye-w-newethguy-boredethguy-2021-logo-1639543419574.png",
        "name": "2021-12-14 - Bullish Radio- The Collector's Eye w/ @newethguy @boredethguy ",
        "tags": Array [
          "poap",
          "event",
        ],
        "year": 2021,
      },
      "rawTokenURI": "0x0000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000002b68747470733a2f2f6170692e706f61702e78797a2f6d657461646174612f31383139392f33313232363833000000000000000000000000000000000000000000",
      "retrievalTime": 0,
      "to": "0xd1811c74361e20280013f3b80c1bc8b9dfe65240",
      "toEthBalance": "0.000000000000000000",
      "tokenIndex": "0x00000000000000000000000000000000000000000000000000000000002fa5fb",
      "tokenURI": "https://api.poap.xyz/metadata/18199/3122683",
      "transactionHash": "0x1c366fda478b01c54a3e590fb07590568b3db2b7c4e8fb3d772d6d34be44d4b8",
    },
    "time": 1639586730000,
    "type": "nft",
  },
  Object {
    "body": Object {
      "blockHash": "0xd14fd6dc5d938c5aaf7cea8b6a1990ddc5014ee126caa8aacc332602d01b6803",
      "blockNumber": 19595084,
      "contract": "0x22c1f6050e56d2876009903609a2cc3fef83b415",
      "errorTokenURI": undefined,
      "from": "0x0000000000000000000000000000000000000000",
      "fromEthBalance": "105987609.019647827235641897",
      "metadata": Object {
        "attributes": Array [
          Object {
            "trait_type": "startDate",
            "value": "12-Dec-2021",
          },
          Object {
            "trait_type": "endDate",
            "value": "12-Dec-2021",
          },
          Object {
            "trait_type": "virtualEvent",
            "value": "true",
          },
          Object {
            "trait_type": "city",
            "value": "",
          },
          Object {
            "trait_type": "country",
            "value": "",
          },
          Object {
            "trait_type": "eventURL",
            "value": "https://app.pwn.finance/",
          },
        ],
        "description": "This POAP gives you access to the PWN lending platform during the closed beta period. 

To access the app: 
1. go to https://app.pwn.finance/
2. connect a wallet which holds this POAP

Happy lending & borrowing!",
        "external_url": "https://api.poap.xyz/metadata/17152/3122684",
        "home_url": "https://app.poap.xyz/token/3122684",
        "image_url": "https://assets.poap.xyz/pwn-app-access-code-2021-logo-1638977620335.png",
        "name": "PWN app access code",
        "tags": Array [
          "poap",
          "event",
        ],
        "year": 2021,
      },
      "rawTokenURI": "0x0000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000002b68747470733a2f2f6170692e706f61702e78797a2f6d657461646174612f31373135322f33313232363834000000000000000000000000000000000000000000",
      "retrievalTime": 0,
      "to": "0xd3ed2b8756b942c98c851722f3bd507a17b4745f",
      "toEthBalance": "0.000000000000000000",
      "tokenIndex": "0x00000000000000000000000000000000000000000000000000000000002fa5fc",
      "tokenURI": "https://api.poap.xyz/metadata/17152/3122684",
      "transactionHash": "0x973844233339a3d7e3d4812983c0e6b12275f119083b37e53fee38b889c49f99",
    },
    "time": 1639586755000,
    "type": "nft",
  },
  Object {
    "body": Object {
      "blockHash": "0x6dd832a0b553e70f7f06c8c1b15fd2d7121b2f3147227d9c19188559890ad666",
      "blockNumber": 19595088,
      "contract": "0x22c1f6050e56d2876009903609a2cc3fef83b415",
      "errorTokenURI": undefined,
      "from": "0x0000000000000000000000000000000000000000",
      "fromEthBalance": "105987609.019647827235641897",
      "metadata": Object {
        "attributes": Array [
          Object {
            "trait_type": "startDate",
            "value": "14-Dec-2021",
          },
          Object {
            "trait_type": "endDate",
            "value": "14-Dec-2021",
          },
          Object {
            "trait_type": "virtualEvent",
            "value": "true",
          },
          Object {
            "trait_type": "city",
            "value": "",
          },
          Object {
            "trait_type": "country",
            "value": "",
          },
          Object {
            "trait_type": "eventURL",
            "value": "https://www.vjdeliria.com/nftribu",
          },
        ],
        "description": "Vj Deliria presents: 

NFTribu, Latin American Art

* 12-DIC-21 *   /    5pm CDT 

-Closing reception - Twitter Spaces & @ Decentraland

-Vj Deliria Events --> https://www.vjdeliria.com/nftribu

-Live Twitter Spaces, Art Gallery Opening --> https://twitter.com/VjDeliria/status/1463904532619014152?s=20

NFTribu is a metaverse exhibition representing Pre-Columbian, Indigenous and Native communities from Latin America through NFT art.
",
        "external_url": "https://api.poap.xyz/metadata/18204/3122685",
        "home_url": "https://app.poap.xyz/token/3122685",
        "image_url": "https://assets.poap.xyz/nftribu-25-latin-american-art-decentraland-12-dic-21-2021-logo-1639545661025.png",
        "name": "NFTribu 2.5,  Latin American Art @ Decentraland  * 12-DIC-21 * ",
        "tags": Array [
          "poap",
          "event",
        ],
        "year": 2021,
      },
      "rawTokenURI": "0x0000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000002b68747470733a2f2f6170692e706f61702e78797a2f6d657461646174612f31383230342f33313232363835000000000000000000000000000000000000000000",
      "retrievalTime": 0,
      "to": "0x9309453b990c46c2e8f3f9c53fc53d530e76059a",
      "toEthBalance": "0.000000000000000000",
      "tokenIndex": "0x00000000000000000000000000000000000000000000000000000000002fa5fd",
      "tokenURI": "https://api.poap.xyz/metadata/18204/3122685",
      "transactionHash": "0xc723131907d9da3b9ff5ef7e82ee9f34e08a9eabc8b5f30d5556d7b9cb59e0fe",
    },
    "time": 1639586775000,
    "type": "nft",
  },
  Object {
    "body": Object {
      "blockHash": "0x0933c78a34167f6f3f1cd6412d2bab5d0e0eece7e43ebd039f88ea74887b67bd",
      "blockNumber": 19595091,
      "contract": "0x22c1f6050e56d2876009903609a2cc3fef83b415",
      "errorTokenURI": undefined,
      "from": "0x0000000000000000000000000000000000000000",
      "fromEthBalance": "105987609.019647827235641897",
      "metadata": Object {
        "attributes": Array [
          Object {
            "trait_type": "startDate",
            "value": "02-Dec-2021",
          },
          Object {
            "trait_type": "endDate",
            "value": "02-Dec-2021",
          },
          Object {
            "trait_type": "virtualEvent",
            "value": "true",
          },
          Object {
            "trait_type": "city",
            "value": "",
          },
          Object {
            "trait_type": "country",
            "value": "",
          },
          Object {
            "trait_type": "eventURL",
            "value": "https://thalesmarket.io/",
          },
        ],
        "description": "This POAP commemorates the first edition of Thales Royale.

On December 2nd, 2021, 2007 players registered in the first edition of Thales Royale. The competition was hosted on Kovan Optimism throughout the following week, and after 6 rounds of intense up or down... only 8 players survived. 

This POAP has been issued to commemorate all players who have participated in the first round, alive or not. 

Will you take the next invitation? Good luck anon. May the luckiest player win.

https://thalesmarket.io/royale?page=scoreboard",
        "external_url": "https://api.poap.xyz/metadata/14512/3122686",
        "home_url": "https://app.poap.xyz/token/3122686",
        "image_url": "https://assets.poap.xyz/thales-royale-poap-2021-logo-1639510478849.png",
        "name": "Thales Royale  POAP",
        "tags": Array [
          "poap",
          "event",
        ],
        "year": 2021,
      },
      "rawTokenURI": "0x0000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000002b68747470733a2f2f6170692e706f61702e78797a2f6d657461646174612f31343531322f33313232363836000000000000000000000000000000000000000000",
      "retrievalTime": 0,
      "to": "0x85447824d5dc10b9fb75928f8104890eb54e7ce1",
      "toEthBalance": "0.000000000000000000",
      "tokenIndex": "0x00000000000000000000000000000000000000000000000000000000002fa5fe",
      "tokenURI": "https://api.poap.xyz/metadata/14512/3122686",
      "transactionHash": "0x7822e51312f9d1e57def7c535ca03e9ed8d602d7eac31988d3f39f56d92fa627",
    },
    "time": 1639586790000,
    "type": "nft",
  },
]
`;
